libmount/src/tab_parse.c

```diff
@@ -673,7 +681,25 @@ static int kernel_fs_postparse(struct libmnt_table *tb,
 	if (src && strcmp(src, "/dev/root") == 0) {
 		char *real = NULL;
 
-		rc = mnt_guess_system_root(fs->devno, tb->cache, &real);
+		/* We will only call mnt_guess_system_root() if it has not
+		 * been called before. Inside a container, mountinfo can contain
+		 * many lines with /dev/root.
+		 */
+		if (saveguess->rc == 0 && saveguess->real == NULL) {
+			rc = mnt_guess_system_root(fs->devno, tb->cache, &real);
+			if (rc == 0 && real != NULL)
+				saveguess->real = strdup(real);
```

Comment:
There is a memory leak, the *real is already allocated string, you do not need strdup() here.
```C
rc = mnt_guess_system_root(fs->devno, tb->cache, &real);
saveguess->real = real;
saveguess->rc = rc;
```

Reply:
Actually, I think it wasn't a memory leak because it would only be executed once and it would be freed by the free statement I had added in  mnt_table_parse_stream().

However, I agree that it's probably unnecessary.  So, I have removed both this strdup() and the free() statement.

When I wrote the code, I was worried that __mnt_fs_set_source_ptr() later in the function would transfer the ownership of the pointer to some other code that could free it before we would re-use it. However, I now understand that this is not the case, the freeing will probably happen later, after mnt_table_parse_stream() is done.